### ISSUE TITLE
spec.py _replace_with dictionary error

### ISSUE THREAD

**Issue by sarich** [spack/spack#2162]
From what I can tell, 'mpi' is getting replaced with my concrete choice 'mpich@3' and so 'mpi' is removed from the dependencies list. But this happens for at least three different packages (hypre, superlu-dist, hdf5), and 'mpi' is no longer in the dictionary.

```
sarich@cetuslac1 spack]$ ./bin/spack spec petsc
Input spec
------------------------------
  petsc

Normalized
------------------------------
  petsc
      ^blas
      ^lapack
      ^python@2.6:2.7
          ^bzip2
          ^ncurses
          ^openssl
              ^zlib
          ^readline
          ^sqlite

Concretized
------------------------------
Traceback (most recent call last):
  File "./bin/spack", line 202, in <module>
    main()
  File "./bin/spack", line 179, in main
    return_val = command(parser, args)
  File "/gpfs/mira-home/sarich/spack/lib/spack/spack/cmd/spec.py", line 57, in spec
    spec.concretize()
  File "/gpfs/mira-home/sarich/spack/lib/spack/spack/spec.py", line 1329, in concretize
    self._expand_virtual_packages(),
  File "/gpfs/mira-home/sarich/spack/lib/spack/spack/spec.py", line 1295, in _expand_virtual_packages
    spec._replace_with(replacement)
  File "/gpfs/mira-home/sarich/spack/lib/spack/spack/spec.py", line 1180, in _replace_with
    del dependent._dependencies[self.name]
KeyError: 'mpi'
```

My packages.yaml file:

```
packages:
  mpich@3:
    paths:
      mpich@3%gcc@4.4.7: /soft/compilers/wrappers/gcc
    buildable: False

  netcdf:
    paths:
      netcdf@4.3.3%gcc@4.4.7: /soft/libraries/netcdf/4.3.3-f4.4.1/cnk-gcc/V1R2M2-20150514
    buildable: False

  hdf5:
    paths:
      hdf5@1.8.14%gcc@4.4.7: /soft/libraries/hdf5/1.8.14/cnk-gcc/V1R2M2-20150515
    buildable: False


  all:
    compiler: [gcc@4.4.7]
    providers:
      mpi: [mpich]
      blas: [openblas]
```

**Comment by sarich** [comment]
I bypassed this del with an existence check, and I'm not getting any python failures anymore. But now I'm getting problem with spack trying to install a new version of mpich instead of using the one in the packages.yaml file. Is there something wrong with it? I'm having some trouble understanding what to do in this file.

**Comment by BarrySmith** [comment]
packages:
  mpich@3:

I think you should have just mpich here, not mpich@3

**Comment by sarich** [comment]
Yep, changed that it now moves forward, but fails on hypre.
The mpich compiler I want to use is 
```
/soft/compilers/wrappers/gcc/mpicc
```

hypre is trying to compile with 
```
/soft/compilers/wrappers/gcc/bin/mpicc
```

In hypre/package.py
```
   def install(self, spec, prefix):
        os.environ['CC'] = spec['mpi'].mpicc
        os.environ['CXX'] = spec['mpi'].mpicxx
        os.environ['F77'] = spec['mpi'].mpif77 <-- Unrelated note: hypre expects 'FC', not 'F77'
```

packages.yaml file now contains:
```
packages:
  mpich:
    paths:
      mpich@3%gcc@4.4.7: /soft/compilers/wrappers/gcc
    buildable: False
```

**Comment by BarrySmith** [comment]
For executables in packages I think spack should not take a "path" but should take the entire name of the executable including its full path.

**Comment by tgamblin** [comment]
Why do you need to use the compiler wrapper here?  Where's `mpicc` actually installed on the ANL machines?  `/soft/compilers/wrappers/gcc` is not the installation directory of MPI, and if we use that as an external it's not going to look like a standard MPI install to Spack.

I realize that ANL has wrappers to make things convenient for users, but I don't think Spack should have to support those, as their behavior varies wildly from facility to facility.

**Comment by tgamblin** [comment]
@wscullin: do you have any insight into this?

**Comment by tgamblin** [comment]
Also @adamjstewart

**Comment by BarrySmith** [comment]
@tgamblin ANL uses the soft system, it is a bit like your "spack compiler wrapper" :-).  The user calls the single compiler wrapper and based on the environment the wrapper calls the appropriate underlying compiler. The user never sees nor needs to see the underlying compiler locations, which depend on exactly what compilers and features you have registered.

>  I realize that ANL has wrappers to make things convenient for users, but I don't think Spack should have to support those, as their behavior varies wildly from facility to facility.

This is sort of like you saying that though "spack" can use (and does) special compiler wrappers, but you are special because others cannot use compiler wrappers also. The issue is that you made a huge assumption about MPI installs that one has a prefix with a  bin directory underneath with the compiler wrappers and a lib with library files etc. There is no standard for MPI that says an MPI install has to support any of the directory structure, in particular that the executables must be in a bin directory.

**Comment by adamjstewart** [comment]
I'm not sure how ALCF does things, but LCRC doesn't use compiler wrappers. At this point, 90% of our installs (including all of our compilers) are done with Spack and aside from a few hiccups (changing hashes), everything is going fine.

**Comment by tgamblin** [comment]
@BarrySmith: Spack has to support both its own installed versions of MPI *and* MPI installations at facilities.  I don't think it's a *huge* assumption to assume that MPI will be installed the way it was intended to be installed by the vendor, especially when the "vendor" of this particular MPI is ANL.

Also, we're not precluding wrappers -- Spack works fine with the MPI compiler wrappers as installed by MPICH an OpenMPI's own installation systems.  What we don't work with and what I don't want to support is nonstandard installation layouts where every facility can invent their own.  At that point I don't think it's MPICH anymore -- it's something else the facility made that they should really upstream if they intend for it to be used by lots of people.

That said, I realize that HPC is an insular community with a lot of weird stuff, so maybe we need support for `soft` the same way we support modules (maybe).  But in this instance I suspect that the compiler wrapper in `/soft/compilers/wrappers/mpicc` actually points at some reasonable `mpicc` someplace else, and I suspect this would work if you just used the real MPI prefix.

**Comment by pramodk** [comment]
I am not sure if `wrappers` matters...here is my setup on another bg-q:

```bash
    mpich:
        paths:
            mpich@3.2%xl@12.1 arch=bgq-CNK-powerpc: /bgsys/drivers/ppcfloor/comm/xl
```
this `/bgsys/...` path exist on every bg-q and contain wrapper scripts. On `Cetus` at ANL I see following:
```
ll /soft/compilers/wrappers/xl/
total 0
lrwxrwxrwx 1 harms software 44 Jun 17  2013 mpic++ -> /soft/compilers/wrappers/redirect-wrapper.sh
lrwxrwxrwx 1 harms software 44 Jun 17  2013 mpicc -> /soft/compilers/wrappers/redirect-wrapper.sh
....
```
So there are all `mpi` standard wrappers exist. 

For XL compilers on our system I have : 
```bash
    paths:
      cc: /opt/ibmcmp/vacpp/bg/12.1/bin/xlc_r
      cxx: /opt/ibmcmp/vacpp/bg/12.1/bin/xlc++
      f77: /opt/ibmcmp/xlf/bg/14.1/bin/xlf_r
```

On `Cetus ` at ANL, XL compiler installation looks like:
```bash
ls /soft/compilers/ibmcmp-feb2013/vacpp/bg/12.1/bin/
bgc89    bgcc     bgxlC      c89    cc        gxlc      new_install    xlc    xlc++_r
bgc89_r  bgcc_r   bgxlc_r    c89_r  cc_r      gxlc++    resetpdf       xlc++  xlC_r
```

So adding those compilers and mpi wrappers path should work (IMO).

**Comment by tgamblin** [comment]
I guess I didn't see above that this is on a BG/Q.  Thanks @pramodk!

Yes -- if you use the vendor MPI install in `/bgsys/drivers/ppcfloor/comm/xl`, this should work.

**Comment by tgamblin** [comment]
@BarrySmith: The potential larger issue here is with the way Spack treats external packages.  We are *trying* to support them (which is risky TBH, as *so much* can go wrong) because you need them on a lot of HPC machines.

When you write:
```yaml
packages:
  mpich:
    paths:
      mpich@3%gcc@4.4.7: /soft/compilers/wrappers/gcc
    buildable: False
```

Your'e telling Spack that the MPI installed in `/soft/compilers/wrappers/gcc` will indeed satisfy the  `mpich@3%gcc@4.4.7` build spec.  This is not true, since `/soft/compilers/wrappers/gcc` is the `soft` compiler install directory, and it's not how `mpich` would actually install.

I would claim that the layout changes required for `soft` wrappers are really a different build configuration, as you're changing the contract of how you use MPI as a dependency.

In Spack, the contract between MPI and other packages is handled with the `mpicc` variable and friends, which are set up by the various MPI packages.  You *could* tweak the `mpich` package with a variant, say `+anl`, that would set things up properly for an  install like this.  You'd need to mess with `setup_dependent_package` in the `mpich` Spack package, and you'd need to have it set the MPI compiler location conditionally to the `/soft/compilers/wrappers/gcc/mpicc` location when the `+anl` variant is enabled.  Then you could write:

```yaml
packages:
  mpich:
    paths:
      mpich@3%gcc@4.4.7+anl: /soft/compilers/wrappers/gcc
    buildable: False
```

And Spack would know about the difference.  But if we can avoid dealing with this by using the vendor wrappers that come with the IBM driver on BG/Q, that would be ideal.  What do the soft wrappers add?  I thought at ANL they just provided some tmp caching for faster builds -- something Spack does already.

**Comment by sarich** [comment]
Thanks @pramodk, I'll give it a shot (but with gcc instead of xl)
By the way, I'm using your bgq-fix pr #1980, thanks! Is there anything holding that up?

Now hypre configure is trying to run a test problem on the front end, and failing because it can't.
```
configure:4321: /bgsys/drivers/ppcfloor/comm/gcc/bin/mpicc -o conftest    conftest.c  >&5
configure:4325: $? = 0
configure:4332: ./conftest
/bgsys/source/srcV1R2M2.15270/comm/sys/buildtools/pami/common/bgq/BgqPersonality.h:102:
/bgsys/source/srcV1R2M2.15270/comm/sys/buildtools/pami/common/bgq/BgqPersonality.h<102>
./configure: line 4334: 10541 Aborted                 (core dumped) ./conftest$ac_cv_exeext
configure:4336: $? = 134
configure:4343: error: in `/tmp/sarich/spack-stage/spack-stage-UF7cdk/hypre/src':
configure:4345: error: cannot run C compiled programs.
If you meant to cross compile, use `--host'.
```

**Comment by pramodk** [comment]
@tgamblin : 
> What do the soft wrappers add? 

I have been using it but never looked into it.  So the layout looks like this:
```bash
[xxxx@cetuslac1 repos]$ ll /soft/compilers/wrappers/gcc
total 0
lrwxrwxrwx 1 harms software 44 Jun 17  2013 mpic++ -> /soft/compilers/wrappers/redirect-wrapper.sh
lrwxrwxrwx 1 harms software 44 Jun 17  2013 mpicc -> /soft/compilers/wrappers/redirect-wrapper.sh
.....
[xxx@cetuslac1 repos]$ ll /soft/compilers/wrappers/xl
total 0
lrwxrwxrwx 1 harms software 44 Jun 17  2013 mpic++ -> /soft/compilers/wrappers/redirect-wrapper.sh
lrwxrwxrwx 1 harms software 44 Jun 17  2013 mpicc -> /soft/compilers/wrappers/redirect-wrapper.sh
```

So all pointing to `redirect-wrapper.sh` which is trivial (IMO) : 
```bash
ALT_PATH=/soft/perftools/autoperf
WRAPPER_PATH=/bgsys/drivers/ppcfloor/comm

dir=$(dirname $0)
wrapper_name=$(basename $0)
wrapper_type=$(basename $dir)
#echo $wrapper_name $wrapper_type

if [ "$DARSHAN_DISABLE" -o "$AP_DISABLE" ];
then
  ${WRAPPER_PATH}/${wrapper_type}/bin/${wrapper_name} "$@"
else
  ${ALT_PATH}/scripts/${wrapper_type}/${wrapper_name} "$@"
fi
```
So at the end we do use system provided wrappers (`xl` or `gcc`) but there is just extra switch to link to `Darshan` automatically.

**Comment by tgamblin** [comment]
Oh right, I forgot about Darshan.  It should be fine to use the vendor wrappers and to build without Darshan.

**Comment by adamjstewart** [comment]
I believe @wcullin mentioned that he had hacked Spack to support SoftEnv, but I'm not sure if it's worth integrating into Spack. From what I understand, even ALCF and LCRC plan on ditching SoftEnv for Modules or LMod.

**Comment by pramodk** [comment]
@sarich : 

> By the way, I'm using your bgq-fix pr #1980, thanks! Is there anything holding that up?

some changes in concretiser needed, like `build` dependencies should be compiled for front-end etc.  I assume, when time permits, @tgamblin could fix those (I don't have enough understanding of internals).

> Now hypre configure is trying to run a test problem on the front end, and failing because it can't.

I haven't built `hypre` but I assume there should be `build option` to disable configure tests during `cross compiling`.

**Comment by BarrySmith** [comment]
> Your'e telling Spack that the MPI installed in /soft/compilers/wrappers/gcc will indeed satisfy the mpich@3%gcc@4.4.7 build spec. This is not true, since /soft/compilers/wrappers/gcc is the soft compiler install directory, and it's not how mpich would actually install.

@tgamblin I think you are making unneccessary assumptions about the mpich install directory structure which you don't need to and shouldn't. This is actually similar to your assumptions about python directory layout. These tools (mpich, python) do not define a directory layout, they just commonly use the same directory layout so your assumptions will work most of the time (but not always, for example python directory structure on Apple). In looking at mpich/package.py I see you make two assumptions 

1) 
 ```
spack_env.set('MPICC',  join_path(self.prefix.bin, 'mpicc'))  
```

assume the executable is always in a bin directory.

2)
 ```
self.spec.mpicxx_shared_libs = [
            join_path(self.prefix.lib, 'libmpicxx.{0}'.format(dso_suffix)),
            join_path(self.prefix.lib, 'libmpi.{0}'.format(dso_suffix))
        ]
```

Assume that the libraries have the above names and are always in a directory called lib, which is next to the bin directory. But you don't need to make this assumption (and you shouldn't) because you can use mpicxx --show to determine the MPI libraries you need to link 

```
$ mpicxx --show
clang++ -I/usr/local/Cellar/open-mpi/2.0.1/include -L/usr/local/opt/libevent/lib -L/usr/local/Cellar/open-mpi/2.0.1/lib -lmpi 
```

As the designer of a package (spack) that you want lots of people to use on lots of systems you don't have the luxury of telling people "oh you can't use it with this, or with this, or with this because they are not 'standard'" you need to live with what people give you.  So don't make assumptions about directory structures to find things, instead use features of the tools themselve to get the information you want. You end up with code that is less fragile to things like "they installed MPI in the wrong way or with a goofy wrapper" and fewer user bug reports when they try things that are not "standard".

I ran the MPICH configure help

```
./configure --help
`configure' configures MPICH 3.2 to adapt to many kinds of systems.

For better control, use the options below.

Fine tuning of the installation directories:
  --bindir=DIR            user executables [EPREFIX/bin]
  --sbindir=DIR           system admin executables [EPREFIX/sbin]
  --libexecdir=DIR        program executables [EPREFIX/libexec]
  --sysconfdir=DIR        read-only single-machine data [PREFIX/etc]
  --sharedstatedir=DIR    modifiable architecture-independent data [PREFIX/com]
  --localstatedir=DIR     modifiable single-machine data [PREFIX/var]
  --libdir=DIR            object code libraries [EPREFIX/lib]
  --includedir=DIR        C header files [PREFIX/include]
  --oldincludedir=DIR     C header files for non-gcc [/usr/include]
  --datarootdir=DIR       read-only arch.-independent data root [PREFIX/share]
  --datadir=DIR           read-only architecture-independent data [DATAROOTDIR]
  --infodir=DIR           info documentation [DATAROOTDIR/info]
  --localedir=DIR         locale-dependent data [DATAROOTDIR/locale]
  --mandir=DIR            man documentation [DATAROOTDIR/man]
  --docdir=DIR            documentation root [DATAROOTDIR/doc/mpich]
```

Note that its configure has the flexibility to install the libraries in any location the user or package manager wants, they definitely don't need to be in the lib directory next to the bin directory where the mpicxx is. Your Spack can't handle the case where someone selected a different libdir but the PETSc system which is much less sophisticated has no problem with it because we don't make any unnessarcy assumptions about directory layouts, we just have MPICH tell use the information we need.

I agree it is nuts to spend huge amounts of time to support one very uncommon and strange case but when adding  support actually makes your tool no more complicated and removes unneeded restrictions on when your tool works then it is better to remove the unneeded assumptions.

**Comment by tgamblin** [comment]
@BarrySmith: I agree we need better autodetection capabilities.  The complication, I think, is where they should go in the spack build "pipeline".

Suggestion (2) above for setting the libs for `mpich` is a really good one because it's likely to be less maintenance and portable across different versions of the `mpich` package.  Of course, `mpich` didn't always support that, so it may not hold for *really* old versions, or for packages that aren't nice enough to tell me the details like`mpich` does.

I kind of like suggestion (1), but how do I know where the `mpicc` compiler lives if it can be anywhere?

The larger concern I have is about reproducibility, something that's built into Spack (to the extent I believe it can be for HPC systems where we have to rely on external things).  And if you detect a compiler from the environment *within* a particular `spack install` command, you don't have any guarantee about what version of `mpich` it is, so the build is all of a sudden not reproducible.

When you tell PETSc's installer to install PETSc and its deps, you're saying "get this working for my environment right now, using what works".  So you can do things like find the `mpicc` that happens to be in your `PATH`.  When you say `spack install <configuration>` you're saying something different:

1. take this abstract specification for something I want to build
2. turn it into a concrete specification, with all the options set
3. build exactly that

The pipeline is useful for a couple reasons:
1. It greatly simplifies what package authors have to specify because `install()` is a translation task and not a search/query/select task.  The fact that the `spec` DAG is known *before* any of the methods in the package get called means I can query my dependencies with things like `+mpi in spec` or `spec['zlib'].version > '1.4'` or `if spec['mpi'] == 'mpich'`.  Package authors don't have to find their own dependencies -- they just know they're built already and only ask how they are configured.  I think this step is what people struggle with the most with install systems, and it's the part that gets brittle when you put different packages together.
2. (2) above generates a `spec.yaml` file with build provenance that I can use to build the same thing again later, and to know exactly what version a user had when I'm trying to figure out bugs.

The consequence for external packages is that the build `spec` needs to be known in advance, and that is what `packages.yaml` is telling you -- it's a mapping from a spec (build configuration) to where you can find it.  This is manual right now (or we can pull it from defaults for some machines like BG/Q and Cray).

To support auto-detection well, we'd need to put it *between* steps 1 and 2 above, and it would generate entries in `packages.yaml`.  Right now we do something like this for system compilers (`compilers.yaml`) but not for system dependencies.  I'd like to generalize the two things by making compilers into dependencies, adding hooks to packages to do things like autodetection, and _just_ have `packages.yaml` and not both `compilers.yaml` and `packages.yaml`.

I worry about making the `packages.yaml` specification overly complicated, and if we were to support *all* the options (which hardly anyone uses) in the `configure` script you pasted, the specification for that *could* get *really* long.  We could probably handle it with sensible defaults and optional args in `packages.yaml` for locations of all the directories -- we could even modify the `prefix` objects Spack passes so that they can support those things.  But do you want that much information in the build provenance?  I think it can be supported within this model, but I'm not sure we can remove _all_ assumptions and still keep things simple.  Right now Spack does make the assumption that a package is basically a prefix.  We violate that on Cray to some extent in that we allow `modules` in `packages.yaml`, so maybe this isn't a major departure.

What do you think?

**Comment by sarich** [comment]
> 
> ```
> configure:4321: /bgsys/drivers/ppcfloor/comm/gcc/bin/mpicc -o conftest    conftest.c  >&5
> configure:4325: $? = 0
> configure:4332: ./conftest
> /bgsys/source/srcV1R2M2.15270/comm/sys/buildtools/pami/common/bgq/BgqPersonality.h:102:
> /bgsys/source/srcV1R2M2.15270/comm/sys/buildtools/pami/common/bgq/BgqPersonality.h<102>
> ./configure: line 4334: 10541 Aborted                 (core dumped) ./conftest$ac_cv_exeext
> configure:4336: $? = 134
> configure:4343: error: in `/tmp/sarich/spack-stage/spack-stage-UF7cdk/hypre/src':
> configure:4345: error: cannot run C compiled programs.
> If you meant to cross compile, use `--host'.
> ```

This is an environment problem that I'm still trying to figure out. When I configure hypre from the command line, I get:
```
configure:4142: /bgsys/drivers/ppcfloor/comm/gcc/bin/mpicc -v >&5
mpicc for MPICH2 version 1.5
Reading specs from /bgsys/drivers/toolchain/V1R2M2_base-efix014/gnu-linux/lib/gcc/powerpc64-bgq-linux/4.4.7/specs
Target: powerpc64-bgq-linux
Configured with: /bgsys/drivers/V1R2M2/ppc64/toolchain/gnu/gcc-4.4.7/configure --prefix=/bgsys/drivers/toolchain/V1R2M2_base-efix01\
4/gnu-linux --target=powerpc64-bgq-linux --with-pkgversion=BGQ-V1R2M2-150309 --host=powerpc64-linux-gnu --build=powerpc64-linux-gnu\
   --disable-multilib --enable-shared --enable-secureplt --disable-libmudflap --disable-libspp --with-headers=/bgsys/drivers/V1R2M2\
/ppc64/toolchain/gnu/build-powerpc64-bgq-linux/tempglibc-4.4.7-install/include --with-libs=/bgsys/drivers/V1R2M2/ppc64/toolchain/gn\
u/build-powerpc64-bgq-linux/tempglibc-4.4.7-install/lib --with-bin=/bgsys/drivers/V1R2M2/ppc64/toolchain/gnu/build-powerpc64-bgq-li\
nux/tempglibc-4.4.7-install/bin --with-sbin=/bgsys/drivers/V1R2M2/ppc64/toolchain/gnu/build-powerpc64-bgq-linux/tempglibc-4.4.7-ins\
tall/sbin --with-long-double-128 --enable-threads=posix --enable-languages=c,c++,fortran
Thread model: posix
gcc version 4.4.7 (BGQ-V1R2M2-150309)
```

but from spack I get:
```
configure:4142: /bgsys/drivers/ppcfloor/comm/gcc/bin/mpicc -v >&5
mpicc for MPICH2 version 1.5
Using built-in specs.
Target: ppc64-redhat-linux
Configured with: ../configure --prefix=/usr --mandir=/usr/share/man --infodir=/usr/share/info --with-bugurl=http://bugzilla.redhat.\
com/bugzilla --enable-bootstrap --enable-shared --enable-threads=posix --enable-checking=release --with-system-zlib --enable-__cxa_\
atexit --disable-libunwind-exceptions --enable-gnu-unique-object --enable-languages=c,c++,objc,obj-c++,java,fortran,ada --enable-ja\
va-awt=gtk --disable-dssi --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-1.5.0.0/jre --enable-libgcj-multifile --enable-java-maintain\
er-mode --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --disable-libjava-multilib --with-ppl --with-cloog --enable-secureplt --with\
-long-double-128 --with-cpu-32=power4 --with-tune-32=power6 --with-cpu-64=power4 --with-tune-64=power6 --build=ppc64-redhat-linux
Thread model: posix
gcc version 4.4.7 20120313 (Red Hat 4.4.7-4) (GCC)
```

**Comment by pramodk** [comment]
@sarich :  what is output of `spack spec petsc +mpi` ?
also could you paste your `compilers.yaml` ?

**Comment by sarich** [comment]
I don't have a compilers.yaml file

```
[sarich@cetuslac1 bspack]$ ./bin/spack spec petsc +mpi
Input spec
------------------------------
  petsc+mpi

Normalized
------------------------------
  petsc+mpi
      ^blas
      ^lapack
      ^mpi
      ^python@2.6:2.7
          ^bzip2
          ^ncurses
          ^openssl
              ^zlib
          ^readline
          ^sqlite

Concretized
------------------------------
  petsc@3.7.4%gcc@4.4.7+boost~complex~debug+double+hdf5+hypre+metis+mpi+mumps~netcdf+shared+superlu-dist~trilinos arch=bgq-cnk-ppc64
      ^hdf5@1.8.14%gcc@4.4.7+cxx~debug+fortran+mpi+shared~szip~threadsafe arch=bgq-cnk-ppc64
      ^hypre@2.11.1%gcc@4.4.7~internal-superlu+shared arch=bgq-cnk-ppc64
          ^mpich@3%gcc@4.4.7+hydra+pmi+romio~verbs arch=bgq-cnk-ppc64
          ^openblas@0.2.19%gcc@4.4.7+fpic~openmp+shared arch=bgq-cnk-ppc64
      ^metis@5.1.0%gcc@4.4.7~debug~gdb~idx64~real64+shared arch=bgq-cnk-ppc64
          ^cmake@3.6.1%gcc@4.4.7~doc+ncurses+openssl~ownlibs~qt arch=bgq-cnk-ppc64
              ^bzip2@1.0.6%gcc@4.4.7 arch=bgq-cnk-ppc64
              ^curl@7.50.3%gcc@4.4.7 arch=bgq-cnk-ppc64
                  ^openssl@1.0.2j%gcc@4.4.7 arch=bgq-cnk-ppc64
                      ^zlib@1.2.8%gcc@4.4.7 arch=bgq-cnk-ppc64
              ^expat@2.2.0%gcc@4.4.7 arch=bgq-cnk-ppc64
              ^libarchive@3.2.1%gcc@4.4.7 arch=bgq-cnk-ppc64
                  ^libxml2@2.9.4%gcc@4.4.7~python arch=bgq-cnk-ppc64
                      ^xz@5.2.2%gcc@4.4.7 arch=bgq-cnk-ppc64
                  ^lz4@131%gcc@4.4.7 arch=bgq-cnk-ppc64
                  ^lzma@4.32.7%gcc@4.4.7 arch=bgq-cnk-ppc64
                  ^lzo@2.09%gcc@4.4.7 arch=bgq-cnk-ppc64
                  ^nettle@3.2%gcc@4.4.7 arch=bgq-cnk-ppc64
                      ^gmp@6.1.1%gcc@4.4.7 arch=bgq-cnk-ppc64
                          ^m4@1.4.17%gcc@4.4.7+sigsegv arch=bgq-cnk-ppc64
                              ^libsigsegv@2.10%gcc@4.4.7 arch=bgq-cnk-ppc64
              ^ncurses@6.0%gcc@4.4.7 arch=bgq-cnk-ppc64
      ^mumps@5.0.2%gcc@4.4.7+complex+double+float~idx64~metis+mpi~parmetis~ptscotch~scotch+shared arch=bgq-cnk-ppc64
          ^netlib-scalapack@2.0.2%gcc@4.4.7~fpic+shared arch=bgq-cnk-ppc64
      ^parmetis@4.0.3%gcc@4.4.7~debug~gdb+shared arch=bgq-cnk-ppc64
      ^python@2.7.12%gcc@4.4.7~tk~ucs4 arch=bgq-cnk-ppc64
          ^readline@6.3%gcc@4.4.7 arch=bgq-cnk-ppc64
          ^sqlite@3.8.5%gcc@4.4.7 arch=bgq-cnk-ppc64
      ^superlu-dist@5.1.1%gcc@4.4.7 arch=bgq-cnk-ppc64
```

**Comment by pramodk** [comment]
> I don't have a compilers.yaml file

ahh...with `spack find compiler` it will generate one in `~/.spack/bgq/compilers.yaml` and I manually updated it. Mine looks like:

```bash
cat ~/.spack/bgq/compilers.yaml
compilers:
- compiler:
    modules: []
    operating_system: redhat6
    paths:
      cc: /usr/lib64/ccache/gcc
      cxx: /usr/lib64/ccache/g++
      f77: /usr/bin/gfortran
      fc: /usr/bin/gfortran
    spec: gcc@4.4.7
- compiler:
    modules: []
    operating_system: redhat6
    paths:
      cc: /opt/ibmcmp/vacpp/bg/12.1/bin/xlc_r
      cxx: /opt/ibmcmp/vacpp/bg/12.1/bin/xlc++
      f77: /opt/ibmcmp/xlf/bg/14.1/bin/xlf_r
      fc: /opt/ibmcmp/xlf/bg/14.1/bin/xlf2008
    spec: xl@12.1
- compiler:
    modules: []
    operating_system: cnk
    paths:
      cc: /usr/bin/bgxlc_r
      cxx: /usr/bin/bgxlc++
      f77: /usr/bin/bgxlf_r
      fc: /usr/bin/bgxlf2008
    spec: xl@12.1
- compiler:
    modules: []
    operating_system: cnk
    paths:
      cc: /bgsys/drivers/ppcfloor/gnu-linux/bin/powerpc64-bgq-linux-gcc
      cxx: /bgsys/drivers/ppcfloor/gnu-linux/bin/powerpc64-bgq-linux-g++
      f77: /bgsys/drivers/ppcfloor/gnu-linux/bin/powerpc64-bgq-linux-gfortran
      fc: /bgsys/drivers/ppcfloor/gnu-linux/bin/powerpc64-bgq-linux-gfortran
    spec: gcc@4.4.7
```

Note that `operating_system: cnk` and its pointing to GNU compilers for compute node. (in your case spack using for front-end node)

**Comment by sarich** [comment]
Thanks, I'll try that. Currently it does point to the wrong compiler for the compute node kernel:

```
- compiler:
    modules: []
    operating_system: cnk
    paths:
      cc: /usr/bin/gcc
      cxx: /usr/bin/g++
      f77: /usr/bin/gfortran
      fc: /usr/bin/gfortran
    spec: gcc@4.4.7
```

**Comment by tgamblin** [comment]
@sarich: can you print the output of `spack config get compilers` if it's not the same as above?  That will merge all the `compilers.yaml` files from different conflict levels and show you what spack sees.

See the new [docs on how config scopes work](http://spack.readthedocs.io/en/latest/configuration.html#configuration-scopes).

**Comment by pramodk** [comment]
@tgamblin : I created #2220 and that might be the reason.

**Comment by sarich** [comment]
```
[sarich@cetuslac1 bspack]$ ./bin/spack config get compilers
compilers:
- compiler:
    modules: []
    operating_system: redhat6
    paths:
      cc: /usr/bin/gcc
      cxx: /usr/bin/g++
      f77: /usr/bin/gfortran
      fc: /usr/bin/gfortran
    spec: gcc@4.4.7
- compiler:
    modules: []
    operating_system: redhat6
    paths:
      cc: /soft/compilers/ibmcmp-may2016/vac/bg/12.1/bin/xlc
      cxx: /soft/compilers/ibmcmp-may2016/vacpp/bg/12.1/bin/xlC
      f77: /soft/compilers/ibmcmp-may2016/xlf/bg/14.1/bin/xlf
      fc: /soft/compilers/ibmcmp-may2016/xlf/bg/14.1/bin/xlf2003
    spec: xl@12.1
- compiler:
    modules: []
    operating_system: cnk
    paths:
      cc: /bgsys/drivers/ppcfloor/gnu-linux/bin/powerpc64-bgq-linux-gcc
      cxx: /bgsys/drivers/ppcfloor/gnu-linux/bin/powerpc64-bgq-linux-g++
      f77: /bgsys/drivers/ppcfloor/gnu-linux/bin/powerpc64-bgq-linux-gfortran
      fc: /bgsys/drivers/ppcfloor/gnu-linux/bin/powerpc64-bgq-linux-gfortran
    spec: gcc@4.4.7
- compiler:
    modules: []
    operating_system: cnk
    paths:
      cc: /soft/compilers/ibmcmp-may2016/vac/bg/12.1/bin/xlc
      cxx: /soft/compilers/ibmcmp-may2016/vacpp/bg/12.1/bin/xlC
      f77: /soft/compilers/ibmcmp-may2016/xlf/bg/14.1/bin/xlf
      fc: /soft/compilers/ibmcmp-may2016/xlf/bg/14.1/bin/xlf2003
    spec: xl@12.1
```

**Comment by sarich** [comment]
Reopening -- the original problem (removing nonexistent item from dictionary) hasn't been fixed

**Comment by sarich** [comment]
```
diff --git a/lib/spack/spack/spec.py b/lib/spack/spack/spec.py
index 17f206d..984fe71 100644
--- a/lib/spack/spack/spec.py
+++ b/lib/spack/spack/spec.py
@@ -1178,7 +1178,8 @@ class Spec(object):
             deptypes = dep_spec.deptypes
 
             # remove self from all dependents.
-            del dependent._dependencies[self.name]
+            if (self.name in dependent._dependencies):
+                del dependent._dependencies[self.name]
 
             # add the replacement, unless it is already a dep of dependent.
             if concrete.name not in dependent._dependencies:
```

**Comment by alalazo** [comment]
Fixed by #2799
